<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>DADA2 tutorial - Paired-ends (PE) reads</title>

<script src="site_libs/header-attrs-2.10/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Metabarcoding tutorial</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="00-setup.html">
    <span class="fa fa-gear"></span>
     
    Setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Data processing
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="01-PE.html">Paired-ends (PE) reads</a>
    </li>
    <li>
      <a href="02-SE.html">Single-end (SE) reads</a>
    </li>
    <li>
      <a href="03-BD.html">Big Data</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Data analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-cleanup.html">Cleanup</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">DADA2 tutorial - Paired-ends (PE) reads</h1>

</div>


<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('left', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<p><br></p>
<hr />
<div id="organize-folders-and-files" class="section level3">
<h3>Organize folders and files</h3>
<p>First we need to create a folder named <code>project</code> with this structure:</p>
<pre><code>project
  |___data</code></pre>
<p>Now, we can move and move all our <code>.fastq.gz</code> to the folder named <code>data</code>, create a <code>.txt</code> file with the metadata and a empty <code>R</code> file where we will store our code. The project folder will look like this:</p>
<pre><code>project
|___data
|     | file1_R1.fastq.gz
|     | file1_R2.fastq.gz
|     | ....... .fastq.gz
|
|___metadata.txt
|___rcode.R</code></pre>
<p><br></p>
</div>
<div id="download-the-data" class="section level3">
<h3>Download the data</h3>
<p>Normally, you would just copy the data from your long-term storage device to the folder <code>project/data</code>. In this case we will download the data from GitHub.</p>
<p><a href="https://downgit.github.io/#/home?url=https://github.com/amalacrino/files_tutorials/tree/main/WWU_tutorial_1/data">Download this file</a> and unzip it inside the folder <code>project/data</code>.</p>
<p><br></p>
</div>
<div id="create-the-metadata-file" class="section level3">
<h3>Create the metadata file</h3>
<p>Within the project folder, you need to save a file named <code>metadata.txt</code> (tab-separated values). The structure of this file should look like this:</p>
<pre><code>SampleID    Factor1       Factor2   ...   FactorN
S1          Control       Root      ...
S2          Treatment     Root      ...</code></pre>
<p>For this tutorial, you can download one from <a href="https://raw.githubusercontent.com/amalacrino/files_tutorials/main/WWU_tutorial_1/metadata.txt">here</a> and place it inside the <code>project/</code> folder.</p>
<p><br></p>
</div>
<div id="create-a-r-file" class="section level3">
<h3>Create a R file</h3>
<p>The quickest way to do so is to:</p>
<ul>
<li>Open R Studio</li>
<li>Click on <code>File/New File/ R Script</code></li>
<li>Click on <code>File/Save</code></li>
<li>Navigate to your project folder and save it</li>
<li>Close R Studio</li>
<li>Open the file you just saved. In this way R Studio will open the file with that relative path already set, so you do not have to worry about changing your code every time you move the folder or when starting a new project.</li>
</ul>
<p><br></p>
</div>
<div id="load-packages" class="section level3">
<h3>Load packages</h3>
<pre class="r"><code>library(&quot;tidyverse&quot;)
library(&quot;gridExtra&quot;)
library(&quot;dada2&quot;)
library(&quot;phyloseq&quot;)
library(&quot;DECIPHER&quot;)
library(&quot;phangorn&quot;)
library(&quot;ape&quot;)
library(&quot;seqinr&quot;)</code></pre>
<p><br></p>
</div>
<div id="file-preparation" class="section level3">
<h3>File preparation</h3>
<p>First, we specify the number of threads to use when multithreaded functions are available.</p>
<pre class="r"><code>n_cores &lt;- 8</code></pre>
<p>Then we run a list of commands to create some folders and specify paths to keep files organized.</p>
<pre class="r"><code>indir &lt;- &quot;data&quot;
filter_dir &lt;- &#39;fastq_filtered&#39;
outdir &lt;- &#39;ASVs&#39;
tax &lt;- &#39;SILVA&#39;
tax_key &lt;- &#39;SILVA/silva_nr99_v138.1_wSpecies_train_set.fa.gz&#39;
metadata_file &lt;- &#39;metadata.txt&#39;
tree &lt;- &#39;tree.tre&#39;
asv.final.table &lt;- &#39;ASV_table.rds&#39;

dir.create(filter_dir, showWarnings = FALSE, recursive = TRUE)
dir.create(tax, showWarnings = FALSE, recursive = TRUE)
dir.create(outdir, showWarnings = FALSE, recursive = TRUE)</code></pre>
<p>And then we can download the SILVA database (v138) that we will use for taxonomic assignment.</p>
<pre class="r"><code>download.file(&quot;https://zenodo.org/record/4587955/files/silva_nr99_v138.1_wSpecies_train_set.fa.gz&quot;, file.path(tax, &#39;silva_nr99_v138.1_wSpecies_train_set.fa.gz&#39;))</code></pre>
<p>And, finally, load the metadata</p>
<pre class="r"><code>metadata_df &lt;- read.table(file = metadata_file, sep = &quot;\t&quot;, header = TRUE, row.names = 1)</code></pre>
<p><br></p>
</div>
<div id="run-the-dada2-workflow" class="section level3">
<h3>Run the DADA2 workflow</h3>
<p>List forwards and reverse <code>.fastq.gz</code> files, given that files were generated to follow the format: <code>*_R1_001.fastq.gz</code> and <code>*_R2_001.fastq.gz</code>. Then we extract the sample name (assuming filenames have format: <code>SAMPLENAME_R*_001.fastq.gz</code>).</p>
<pre class="r"><code>fnFs &lt;- sort(list.files(indir, pattern=&quot;_R1_001.fastq.gz&quot;, full.names = TRUE))
fnRs &lt;- sort(list.files(indir, pattern=&quot;_R2_001.fastq.gz&quot;, full.names = TRUE))
sampleIDs &lt;- sapply(strsplit(basename(fnFs), &quot;_&quot;), `[`, 1)</code></pre>
<p>Inspect the read quality profiles. First forward.</p>
<pre class="r"><code>plotQualityProfile(fnFs[1])</code></pre>
<p>And, then, reverse.</p>
<pre class="r"><code>plotQualityProfile(fnRs[1:2])</code></pre>
<p>Then we can define the file path for filtered files and define filenames.</p>
<pre class="r"><code>filtFs &lt;- file.path(filter_dir, paste0(sampleIDs, &quot;_F_filt.fastq.gz&quot;))
filtRs &lt;- file.path(filter_dir, paste0(sampleIDs, &quot;_R_filt.fastq.gz&quot;))
names(filtFs) &lt;- sampleIDs
names(filtRs) &lt;- sampleIDs</code></pre>
<p>Filter and trim raw reads.</p>
<pre class="r"><code>filter_results &lt;- filterAndTrim(fnFs, filtFs, fnRs, filtRs,
                                truncLen = c(250,250),
                                rm.phix = FALSE,
                                multithread = n_cores, 
                                compress = FALSE, verbose = TRUE)</code></pre>
<details>
<summary>
  <svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"></path></svg>   <strong>What can be changed?</strong> (Click here)
</summary>
<p><code>compress = TRUE</code> (Optional). Default TRUE. If TRUE, the output fastq file(s) are gzipped.</p>
<p><code>truncQ = 2</code> (Optional). Default 2. Truncate reads at the first instance of a quality score less than or equal to <code>truncQ</code>.</p>
<p><code>truncLen = 0</code> (Optional). Default 0 (no truncation). Truncate reads after <code>truncLen</code> bases. Reads shorter than this are discarded.</p>
<p><code>trimLeft = 0</code> (Optional). Default 0. The number of nucleotides to remove from the start of each read. If both <code>truncLen</code> and <code>trimLeft</code> are provided, filtered reads will have length <code>truncLen</code>-<code>trimLeft</code>.</p>
<p><code>trimRight = 0</code> (Optional). Default 0. The number of nucleotides to remove from the end of each read. If both <code>truncLen</code> and <code>trimRight</code> are provided, truncation will be performed after <code>trimRight</code> is enforced.</p>
<p><code>maxLen = Inf</code> (Optional). Default Inf (no maximum). Remove reads with length greater than <code>maxLen.</code> <code>maxLen</code> is enforced before trimming and truncation.</p>
<p><code>minLen = 20</code> (Optional). Default 20. Remove reads with length less than <code>minLen.</code> <code>minLen</code> is enforced after trimming and truncation.</p>
<p><code>maxN = 0</code> (Optional). Default 0. After truncation, sequences with more than maxN Ns will be discarded. Note that dada does not allow Ns.</p>
<p><code>minQ = 0</code> (Optional). Default 0. After truncation, reads contain a quality score less than minQ will be discarded.</p>
<p><code>maxEE = Inf</code> (Optional). Default <code>Inf</code> (no EE filtering). After truncation, reads with higher than maxEE “expected errors” will be discarded. Expected errors are calculated from the nominal definition of the quality score: EE = sum(10^(-Q/10))</p>
<p><code>rm.phix = TRUE</code> (Optional). Default TRUE. If TRUE, discard reads that match against the phiX genome, as determined by <code>isPhiX</code>.</p>
</details>
<p><br></p>
<p>Check how many reads have been filtered.</p>
<pre class="r"><code>filter_results</code></pre>
<p>Remove those samples that have been completely filtered out.</p>
<pre class="r"><code>exists &lt;- file.exists(fnFs) &amp; file.exists(filtFs)
filtFs &lt;- filtFs[exists]
exists &lt;- file.exists(fnRs) &amp; file.exists(filtRs)
filtRs &lt;- filtRs[exists]</code></pre>
<p>Learn error rates.</p>
<pre class="r"><code>errF &lt;- learnErrors(filtFs, multithread = n_cores, verbose = TRUE)
errR &lt;- learnErrors(filtRs, multithread = n_cores, verbose = TRUE)</code></pre>
<p>Run sample inference.</p>
<pre class="r"><code>dadaFs &lt;- dada(filtFs, err=errF, pool = FALSE, multithread = n_cores)
dadaRs &lt;- dada(filtRs, err=errR, pool = FALSE, multithread = n_cores)</code></pre>
<details>
<summary>
  <svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"></path></svg>   <strong>What can be changed?</strong> (Click here)
</summary>
<p><code>pool</code> (Optional). If <code>pool = TRUE</code>, the algorithm will pool together all samples prior to sample inference. If <code>pool = FALSE</code>, sample inference is performed on each sample individually. If <code>pool = "pseudo"</code>, the algorithm will perform pseudo-pooling between individually processed samples.</p>
<p><code>priors</code> (Optional). The priors argument provides a set of sequences for which there is prior information suggesting they may truly exist, i.e. are not errors. The abundance p-value of dereplicated sequences that exactly match one of the priors are calculated without conditioning on presence, allowing singletons to be detected, and are compared to a reduced threshold ‘OMEGA_P’ when forming new partitions.</p>
</details>
<p><br></p>
<p>Merge paired reads.</p>
<pre class="r"><code>mergers &lt;- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)</code></pre>
<details>
<summary>
  <svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"></path></svg>   <strong>What can be changed?</strong> (Click here)
</summary>
<p><code>minOverlap = 12</code> (Optional). Default 12. The minimum length of the overlap required for merging the forward and reverse reads.</p>
<p><code>maxMismatch = 0</code> (Optional). Default 0. The maximum mismatches allowed in the overlap region.</p>
</details>
<p><br></p>
<p>Construct ASV table.</p>
<pre class="r"><code>seqtab_all &lt;- makeSequenceTable(mergers)</code></pre>
<details>
<summary>
  <svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"></path></svg>   <strong>What can be changed?</strong> (Click here)
</summary>
<p><code>orderBy</code> (Optional). <code>character(1)</code>. Default “abundance”. Specifies how the sequences (columns) of the returned table should be ordered (decreasing). Valid values: “abundance”, “nsamples”, NULL.</p>
</details>
<p><br></p>
<p>Remove chimeras.</p>
<pre class="r"><code>seqtab &lt;- removeBimeraDenovo(seqtab_all,
                             method = &#39;consensus&#39;,
                             multithread = n_cores,
                             verbose = TRUE)</code></pre>
<details>
<summary>
  <svg viewBox="0 0 512 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"></path></svg>   <strong>What can be changed?</strong> (Click here)
</summary>
<p><code>method</code> (Optional). Default is “consensus”. Only has an effect if a sequence table is provided. If <code>pooled</code>: The samples in the sequence table are all pooled together for bimera identification (<code>isBimeraDenovo</code>). If <code>consensus</code>: The samples in a sequence table are independently checked for bimeras, and a consensus decision on each sequence variant is made (<code>isBimeraDenovoTable</code>). If <code>per-sample</code>: The samples in a sequence table are independently checked for bimeras, and sequence variants are removed (zeroed-out) from samples independently (<code>isBimeraDenovo</code>).</p>
</details>
<p><br></p>
<p>Track reads through the pipeline.</p>
<pre class="r"><code>getN &lt;- function(x) sum(getUniques(x))
track &lt;- cbind(filter_results, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab))
colnames(track) &lt;- c(&quot;input&quot;, &quot;filtered&quot;, &quot;denoisedF&quot;, &quot;denoisedR&quot;, &quot;merged&quot;, &quot;nonchim&quot;)
rownames(track) &lt;- sampleIDs
track</code></pre>
<p>Assign taxonomy.</p>
<pre class="r"><code>taxa &lt;- assignTaxonomy(seqtab, tax_key, multithread = n_cores)
colnames(taxa) &lt;- c(&#39;Kingdom&#39;, &#39;Phylum&#39;, &#39;Class&#39;, &#39;Order&#39;, &#39;Family&#39;, &#39;Genus&#39;, &quot;Species&quot;)</code></pre>
<p>Rename ASVs (optional).</p>
<pre class="r"><code>otutable &lt;- seqtab
colnames(otutable) &lt;- paste(&#39;ASV&#39;, 1:ncol(seqtab), sep = &#39;_&#39;)
row.names(taxa) &lt;- paste(&#39;ASV&#39;, 1:ncol(seqtab), sep = &#39;_&#39;)</code></pre>
<p>Generate <code>.fasta</code> file with ASVs.</p>
<pre class="r"><code>asv_seqs &lt;- colnames(seqtab)
asv_headers &lt;- paste(&#39;ASV&#39;, 1:ncol(seqtab), sep = &#39;_&#39;)
asv_fasta &lt;- c(rbind(asv_headers, asv_seqs))
write(asv_fasta, file = file.path(outdir, &#39;ASVs.fa&#39;)) #optional</code></pre>
<p>Align sequences.</p>
<pre class="r"><code>seqs &lt;- getSequences(seqtab)
names(seqs) &lt;- asv_headers 
alignment &lt;- AlignSeqs(DNAStringSet(seqs), anchor=NA, processors = n_cores)
phang.align &lt;- phyDat(as(alignment, &quot;matrix&quot;), type=&quot;DNA&quot;)
write.phyDat(phang.align, file = file.path(outdir, &#39;alignment.fasta&#39;), format=&quot;fasta&quot;) #optional</code></pre>
<p>Build phylogenetic tree with ASVs.</p>
<pre class="r"><code>dna_dist &lt;- dist.ml(phang.align, model=&quot;JC69&quot;)
asv_UPGMA &lt;- upgma(dna_dist)
fit &lt;- pml(asv_UPGMA, phang.align)
fitJC &lt;- optim.pml(fit, model = &quot;JC&quot;, rearrangement = &quot;stochastic&quot;)
tree &lt;- bootstrap.pml(fitJC, bs=1, optNni=TRUE, multicore=TRUE, control = pml.control(trace=0))</code></pre>
<p><br></p>
</div>
<div id="generate-phyloseq-object" class="section level3">
<h3>Generate phyloseq object</h3>
<pre class="r"><code>ps &lt;- phyloseq(otu_table(otutable, taxa_are_rows = FALSE),
                   sample_data(metadata_df),
                   tax_table(taxa),
                   tree[[1]])
ps</code></pre>
<p>Clean environment and save it.</p>
<pre class="r"><code>rm(list=setdiff(ls(), c(&quot;ps&quot;)))
save.image(file = asv.final.table)</code></pre>
<pre class="r"><code>sessionInfo()</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
