<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Data cleanup</title>

<script src="site_libs/header-attrs-2.10/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Metabarcoding tutorial</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="00-setup.html">
    <span class="fa fa-gear"></span>
     
    Setup
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Data processing
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="01-PE.html">Paired-ends (PE) reads</a>
    </li>
    <li>
      <a href="02-SE.html">Single-end (SE) reads</a>
    </li>
    <li>
      <a href="03-BD.html">Big Data</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Data analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-cleanup.html">Cleanup</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Data cleanup</h1>

</div>


<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('left', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<p><br></p>
<hr />
<div id="download-the-data-and-organize-it" class="section level3">
<h3>Download the data and organize it</h3>
<p>First we need to create a folder named <code>project</code> with this structure:</p>
<pre><code>project
  |___analysis.Rmd
  |___ASV_table.rds</code></pre>
<p>The ASV table for this tutorial is available <a href="https://github.com/amalacrino/files_tutorials/blob/main/WWU_tutorial_3/ASV_table.rds">here</a>.</p>
<p><br></p>
</div>
<div id="load-packages" class="section level3">
<h3>Load packages</h3>
<pre class="r"><code>library(&quot;dplyr&quot;)
library(&quot;phyloseq&quot;)
library(&quot;DESeq2&quot;)
library(&quot;decontam&quot;)
library(&quot;microbiome&quot;)
library(&quot;ggplot2&quot;)
library(&quot;data.table&quot;)
library(&quot;limma&quot;)
library(&quot;vegan&quot;)</code></pre>
<p><br></p>
</div>
<div id="load-data" class="section level3">
<h3>Load data</h3>
<pre class="r"><code>load(file = &#39;ASV_table.rds&#39;)</code></pre>
<p><br></p>
</div>
<div id="explore-the-raw-data" class="section level3">
<h3>Explore the raw data</h3>
<pre class="r"><code>metadata &lt;- data.frame(sample_data(ps))
asv.table &lt;- data.frame(t(otu_table(ps)))
tax.table &lt;- data.frame(tax_table(ps))</code></pre>
<p><br></p>
</div>
<div id="remove-plastidial-reads" class="section level3">
<h3>Remove plastidial reads</h3>
<pre class="r"><code>ps &lt;- subset_taxa(ps, Class != &quot;Chloroplast&quot;)
ps &lt;- subset_taxa(ps, Order != &quot;Chloroplast&quot;)
ps &lt;- subset_taxa(ps, Family != &quot;Mitochondria&quot;)</code></pre>
<p><br></p>
</div>
<div id="remove-possible-contaminants" class="section level3">
<h3>Remove possible contaminants</h3>
<pre class="r"><code>sample_data(ps)$is.neg &lt;- sample_data(ps)$Variety == &quot;Control&quot;
contamdf.prev &lt;- isContaminant(ps, method=&quot;prevalence&quot;, neg=&quot;is.neg&quot;, threshold = 0.05)
cont.remove &lt;- subset(contamdf.prev, contaminant == &quot;TRUE&quot;)
cont.remove &lt;- row.names(cont.remove)
allTaxa = taxa_names(ps)
allTaxa &lt;- allTaxa[!(allTaxa %in% cont.remove)]
ps &lt;-  prune_taxa(allTaxa, ps)
ps &lt;- subset_samples(ps, Variety != &quot;Control&quot;)</code></pre>
<p><br></p>
</div>
<div id="remove-singletons" class="section level3">
<h3>Remove singletons</h3>
<pre class="r"><code>ps &lt;- filter_taxa(ps, function (x) {sum(x &gt; 0) &gt; 1}, prune=TRUE)</code></pre>
<p><br></p>
</div>
<div id="remove-samples-with-low-reads" class="section level3">
<h3>Remove samples with low reads</h3>
<pre class="r"><code>ps &lt;- prune_samples(sample_sums(ps) &gt;= 1000, ps)</code></pre>
<p><br></p>
</div>
<div id="create-a-phyloseq-object-with-normalized-data" class="section level3">
<h3>Create a phyloseq object with normalized data</h3>
<pre class="r"><code>GM &lt;- ps
otu_table(GM) &lt;- otu_table(GM)+1
diagdds = phyloseq_to_deseq2(GM, ~ 1)
ts = counts(diagdds)
geoMeans = apply(ts, 1, function(row) if (all(row == 0)) 0 else exp(mean(log(row[row != 0]))))
diagdds = estimateSizeFactors(diagdds, geoMeans=geoMeans)
diagdds = estimateDispersions(diagdds)
diagvst = getVarianceStabilizedData(diagdds)
diagdds.c &lt;- removeBatchEffect(diagvst)
diagdds.c[diagdds.c&lt;0] &lt;- 0
ps.norm &lt;- ps
otu_table(ps.norm) &lt;- otu_table(diagdds.c, taxa_are_rows = TRUE)
ps.norm</code></pre>
<p><br></p>
</div>
<div id="check-sequencing-depth" class="section level3">
<h3>Check sequencing depth</h3>
<pre class="r"><code>sdt = data.table(as(sample_data(ps), &quot;data.frame&quot;),
                 TotalReads = sample_sums(ps), keep.rownames = TRUE)
setnames(sdt, &quot;rn&quot;, &quot;SampleID&quot;)
sdt</code></pre>
<p><br></p>
</div>
<div id="print-the-rarefaction-curve" class="section level3">
<h3>Print the rarefaction curve</h3>
<pre class="r"><code>rarecurve(otu_table(ps), step=200, cex=0.5, label = F)</code></pre>
<p><br></p>
</div>
<div id="check-how-many-samples-are-left-after-filtering" class="section level3">
<h3>Check how many samples are left after filtering</h3>
<pre class="r"><code>sample_data(ps) %&gt;%
  group_by(Sample_type, Variety) %&gt;%
  summarise(nr_Samples = n())</code></pre>
<p><br></p>
</div>
<div id="plot-relative-abundance" class="section level3">
<h3>Plot relative abundance</h3>
<pre class="r"><code>glom &lt;- microbiome::aggregate_taxa(ps.norm, &quot;Genus&quot;)
glom &lt;- microbiome::transform(glom, &quot;compositional&quot;)
dat &lt;- psmelt(glom)
filt.gen &lt;- dat %&gt;% group_by(Genus) %&gt;% summarize(mean = mean(Abundance)) %&gt;% filter(mean &lt;= 0.01)
dat &lt;- subset(dat, !(OTU %in% filt.gen$Genus))
dat &lt;- dat %&gt;% group_by(Sample_type, Genus) %&gt;% summarize(cs = mean(Abundance)) %&gt;% mutate(cs = cs/sum(cs)) 

taxa_plot &lt;- ggplot(dat, aes(x = Sample_type, y = cs, fill = Genus)) +
                          theme_bw(base_size = 14) +
                          geom_bar(stat=&quot;identity&quot;) +
                          labs(y = &quot;Relative proportion&quot;) +
                          theme(axis.title.x=element_blank(),
                                axis.text.x = element_text(color=&quot;black&quot;, angle = 45, hjust = 1),
                                axis.text.y = element_text(color=&quot;black&quot;),
                                axis.title.y = element_text(color=&quot;black&quot;),
                                panel.grid = element_blank(),
                                strip.background = element_blank(),
                                strip.text = element_blank()) +
                          scale_y_continuous(labels = function(x) paste0(x*100, &quot;%&quot;))
taxa_plot</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
